content/en/blog/general/fuzzing-audit/index.md

```diff
@@ -0,0 +1,35 @@
+
+Read the full report for the audit here: [Kyverno Fuzzing Security Audit Report](kyverno-2023-fuzzing-security-audit.pdf).
+
+## Fuzzing
+
+Fuzzing is a technique for finding software bugs in an automated manner. In Go, this includes bugs such as nil-dereference, out-of-bounds panics, out-of-range panics, out-of-memory panics as well as logic bugs. To test your software by way of fuzzing, you need two things:
+
+1. A fuzzing engine
+2. A test harness
+
+The fuzzing engine generates the input test cases that it makes available in the test harness. It is the user's job to set up the harness such that it passes the test cases on to the API that the harness is meant to test. Kyverno is implemented in Go, and the auditors wrote the fuzz tests using [the standard library fuzzing engine](https://go.dev/security/fuzz/). Once the user has written a harness and runs it, the fuzzer runs indefinitely until it detects a crash or the user stops it. Each time the harness runs (each fuzz iteration), the engine will generate a new test case by mutating over a collection of test cases that explore new code. This collection of test cases is called a corpus.
+
+## Kyverno fuzzers
+
+All of the fuzzers from the audit are hosted in Kyvernos repository in the package directory that they test. Ada Logics assessed the Kyverno code base by way of static callgraph analysis to determine the optimal entrypoints for the fuzzers and found that the validation routines for Validate, Mutate and Verify Image rules reach a large part of the Kyverno source tree. They then wrote a fuzzer for each routine which passes both a valid Kyverno Cluster Policy as well pseduo-random unstructured Kubernetes resource, testing these validation routines for both reliability and security issues. One of the issues that the fuzzers found during the audit was triggered by Kyvernos mishandling of the input resource. The Validate, Mutate and Verify Image fuzzers require a substantial setup depending on multiple mocked types, and Ada Logics did several iterations to improve their effectiveness. This included several code optimizations in the fuzzers including how the fuzzers generate policy rules and the Kubernetes resources before passing them onto the entrypoints. Another optimization was to add a dictionary; Native Go fuzzing does not support dictionaries, but OSS-Fuzz does when running the fuzzers. A dictionary is a list of tokens that the fuzzing engine will mutate over in the first iterations when starting to run a fuzzer, and it can be a great resource to get a fuzzer started with meaningful test cases. In Kyvernos case, Ada Logics scraped the json tags of several Kubernetes resource types so that the fuzzers can use these when generating the Kubernetes resources.
```

Comment:
"Kyverno's"

Reply:
"JSON tags"

Reply:
Add link to https://github.com/kyverno/kyverno/tree/main/test/fuzz?